registry/storage/driver/s3-aws/s3.go

```diff
@@ -942,15 +942,152 @@ func (d *driver) Walk(ctx context.Context, from string, f storagedriver.WalkFn)
 	return nil
 }
 
-func (d *driver) doWalk(parentCtx context.Context, objectCount *int64, path, prefix string, f storagedriver.WalkFn) error {
+func (d *driver) doWalk(parentCtx context.Context, objectCount *int64, path, prefix string, f storagedriver.WalkFn, opts ...storagedriver.WalkOpt) error {
+	options := storagedriver.ToWalkOptions(opts...)
+	if options.StopOnErrSkipDir {
+		return d.doWalkWithoutSkipDir(parentCtx, objectCount, path, prefix, f)
+	}
+	_, err := d.doWalkRecursive(parentCtx, objectCount, path, prefix, f)
+	return err
+}
+
+type walkInfoContainer struct {
+	storagedriver.FileInfoFields
+	prefix *string
+}
+
+// Path provides the full path of the target of this file info.
+func (wi walkInfoContainer) Path() string {
+	return wi.FileInfoFields.Path
+}
+
+// Size returns current length in bytes of the file. The return value can
+// be used to write to the end of the file at path. The value is
+// meaningless if IsDir returns true.
+func (wi walkInfoContainer) Size() int64 {
+	return wi.FileInfoFields.Size
+}
+
+// ModTime returns the modification time for the file. For backends that
+// don't have a modification time, the creation time should be returned.
+func (wi walkInfoContainer) ModTime() time.Time {
+	return wi.FileInfoFields.ModTime
+}
+
+// IsDir returns true if the path is a directory.
+func (wi walkInfoContainer) IsDir() bool {
+	return wi.FileInfoFields.IsDir
+}
+
+// doWalkRecursive is an implementation of Walk that is optimized for listing high level directories, such as repositories,
+// where storagedriver.ErrSkipDir is utilized to skip over large numbers of keys.
+func (d *driver) doWalkRecursive(parentCtx context.Context, objectCount *int64, path, prefix string, f storagedriver.WalkFn) (bool, error) {
```

Comment:
`doWalkRecursive` is just the walk implementation prior to https://github.com/distribution/distribution/pull/3480